### DISCUSSION TITLE
Oqtane with EF Core In-memory Provider?

### DISCUSSION THREAD

**Discussion by RodDaSilva** [oqtane/oqtane.framework#2335] [General]
Hi!

Has anyone tried Oqtane with the EF Core In-memory Provider?  I am wondering because I want to try to make Oqtane work with a custom data source via EF Core.  There isn't a lot of info / documentation on how to write  a custom EF Core Data Provider.  I therefore thought that if I can get Oqtane working with the In-memory EF Core Data Provider Microsoft provides (with full source code), I could use it as a starting point.  I haven't looked at Oqtane in over a year, but I seem to remember that it didn't use store procs and its relational table access was pretty straight forward.   Therefore I was wondering if a simple In-Memory Provider would handle them?

Just wondering if anyone has perhaps used the In-memory EF Core Data Provider for say unit testing Oqtane?

Thanks,
Rod

**Comment by hishamco** [comment]
Right, In-memory provider is suited for unit tests. Regarding to to build a custom provider I suggest to follow what Oqtane did for the current Database Providers

**Comment by RodDaSilva** [comment]
What do you mean?  What did Oqtane do regarding custom Database Providers?  Did the Oqtane development group create a custom Database Provider?  I had understand they simply standardized on EF Core and made use of existing Database Providers.   I can find only outdated documentation from the EF 3.0 era for developing custom Database Providers, which are all out of date for EF Core 6/7.   My thinking is that if I can at least get Oqtaine to install and run using the current EF Core 6/7 In-memory Database Provider, then I can start from there and modify the code to work with my custom data source.   I am just wondering if anyone has ran unit tests with Oqtane on this In-Memory Database Provider successfully, or does Oqtane use advanced "relational provider" features that wouldn't work with the In-Memory provider?

**Comment by hishamco** [comment]
Could you please describe what kind of provider you are working on?

**Comment by RodDaSilva** [comment]
I have a proprietary data source that I want to back Oqtane with.  This data source is a decentralized database technology.  It is part of the World Computer project ([WorldComputer.org](worldcomputer.org) for more info).  The big idea is to store the data for an Oqtane instance in a **decentralized** data store rather than in a traditional centralized data store.  

I would rather NOT use EF Core for data access but my understand is that it is "baked in" to Oqtane.  At one time there was a data "repository" abstraction which would be great to tie into but, I suspect all of the data "access" logic in Oqtane is likely tied to LINQ and EF. So to achieve what I want to do I likely need to develop a custom EF Core 6/7 Data Provider to get Oqtane to store its data to my decentralized data store.

My decentralized data store is not SQL based.  But it is pseudo relational.  It has "tables" with "indexes" (orders) and the tables can be related with "relations".   If you are old school its very similar to the old DBF standard from Ashton Tate from the 80's which was very popular back in the day - aka DOS application development (Ashton Tate had 18 million registered users at one time!).   DBF files (tables) and their indexes where everywhere back then (pre DOS networks!!) where an app simply ran on one computer, because they were fast and efficient at storing a **single user** application's data locally.   But when networks arrived  and the apps began being shared by many users simultaneously over networks, sharing this "local data" became hard.   Hence the rise of centralized RDBMS (aka SQL databases) took over.   

However, my thesis for the world computer project is there is only **one** Computer, and all apps and users share a logically "local" massive virtual drive attached to this one massive computer.   In this vision there is no concept of a network for either the user or the application developer developing applications for this computer.   It would be like you are teleported back to the early 1980s in terms of how simple it would be to develop applications for this computer (networks exponentially complicate software design).   If can imagine this then you will understand that it would allow the pendulum to swing back away from centralized data repositories toward the faster and more simply (logically) "local" data storage fro each app.  Again check out WorldComputer.org for a better understanding of the entire vision (including 3 whitepapers).   

I was a DNN user for nearly 10 years, and the World Computer project is entirely developed in C#.  I am also a big fan of Blazor and so want to target the project's front-end in Blazor.   Hence its natural that I embrace the Oqtane  project to create the base UI layer of a "single computer-abstraction world computer" (i.e.; a computer built out of P2P nodes with no central servers).  

Thus I arrive at the need for a custom EF Core data provider that will connect Oqtane to a decentralized (i.e.; P2P) data store.  The problem is there has been lots of "[churn](https://docs.microsoft.com/en-us/ef/core/providers/provider-log)" on the EF Data Provider APIs and the documentation for creating a data provider for the current version of EF Core (6/7) is woefully out of data.  The closest thing to what I am looking for is a "[File-based](https://github.com/morrisjdev/FileContextCore)"  EF Data Provider but it only works with DotNet Core 3.0 and won't compile in .NET 6.0 / EF Core 6 due to the many API breaking changes (i.e.; the churn mentioned above) since then.

The simplest starting point that I have found for creating a custom EF Core 6/7 Data Provider is the Microsoft maintained "[In-Memory](https://github.com/dotnet/efcore/tree/main/src/EFCore.InMemory)" Data Provider that is designed for unit testing, etc.,.   If Oqtane will install and run on this data provider, I could begin there to modify the code to point the provider to my custom decentralized data source which would have a lot in common (semantically speaking) with an "in-memory" model.  I posted my original question here in case anyone in the know would be able to say "no it won't work because Oqtane assumes its data providers needs X feature which the in-memory provide doesn't support" - thus saving me a bunch of time researching this.

I hope that provides the background for my original question.

**Comment by sbwalker** [comment]
Oqtane does have a logical Repository layer whose responsibility is purely related to data access. The Repository layer currently uses EF Core for all data access and migrations. The Repository layer itself is not designed to be dynamically replaced at runtime, but it utilizes a custom abstraction which allows it to meet the multi-tenant needs of the framework and support multiple databases (SQL Server, mySQL, SQLite, and PostgreSQL). Different databases are supported through a custom provider model which relies on an IDatabase interface and DatabaseBase base class (see https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Database.Sqlite/SqliteDatabase.cs). Please note that all of the currently supported databases are RDMS's - as Oqtane does rely on referential integrity and behaviors such as cascade delete.

**Comment by RodDaSilvaWCO** [comment]
@sbwalker  - thanks for the response. 

I have dug a bit deeper on this and have determined that the EF Core In-Memory Data Provider supports referential integrity and cascading deletes so that's not a problem.   The biggest issue here will be support for **Migrations**, which the In-Memory provider does not support (nor does any **non-relational** provider at the moment).  However, I am looking into ways to get around this.   I don't see a technical reason why a non-relational provider couldn't be made to support Migrations.

My idea is that if I can make Oqtane's Migrations work with the In-Memory provider by tweaking the provider itself (i.e. a custom In-Memory provider)  which would then provide a pattern from which to tackle a more useful **non-relational** data store such as files on the file system.

I will play with this some more.

Thanks.

**Reply by hishamco** [reply]
If I'm not wrong EF Core already supports Non SQL databases such as Cosmos, you can have a look to it

**Reply by sbwalker** [reply]
It makes sense that the In-Memory provider does not support migrations. Migrations are needed for transforming data from one physical state to another... but the In-Memory provider has no physical state - it is purely virtual.

**Reply by RodDaSilvaWCO** [reply]
> It makes sense that the In-Memory provider does not support migrations. Migrations are needed for transforming data from one physical state to another... but the In-Memory provider has no physical state - it is purely virtual.

I don't agree.   Any database (in-memory or not) has a "logical" schema that defines the data's structure.  As such that schema can change (i.e.; need versioning).    The fact that EF Core only currently supports Migrations for non-Relational data providers is a short coming in EF Core IMO.   SQL is a standard and they did a lot of work to create a Migration facility to generate SQL Data Definition Language (DDL) scripts to move the schema version of a SQL database Up or Down including the ability to customize that SQL script generation for a particular flavor of SQL.   

However, I might have my own DDL for my custom database store format, and my need to move my database schema Up or Down is no less important, and so the concept of Migrations is equally valid for non-relational data providers as well.  Note that the issue isn't what medium the data is stored in (i.e.; on disk or in-memory) but rather that they simply don't provide a default facility for Migrations for non-relational data providers.   There should be a facility for me to override completely the generation of custom DDL scripts for my custom database using the Migration facility.  However, they do not enable it (out-of-the-box anyway) unless you implement a Relational database provider.

However, upon cursory inspection of the EF Core source code, it appears that all of the Migration facilities are implemented as DI services.  They simply aren't included in the non-Relational EntityFrameworkServicesBuilder, where as they are added to the EntityFrameworkRelationalServicesBuilder which inherits from EntityFrameworkServicesBuilder.  All non-relational data providers only use EntityFrameworkServicesBuilder.  So in theory it should be possible to add back the Migration services into a custom subclass of EntityFrameworkServicesBuilder to achieve the required.

I am going to throw some more time at this as I would really like to prove the case that Oqtane can work with custom non-relational data stores (including Migrations).   I am only starting with In-Memory because the source code is all there ready to customize.

**Comment by RodDaSilvaWCO** [comment]
Yes.  But the "Migrations" feature is only currently available for Relational data providers.
